Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,1150 @@
+new file mode 100644
+index 000000000..b744ae926
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/util/PlayerTicketAreaMap.java
+@@ -0,0 +1,395 @@
++package com.destroystokyo.paper.util;
++
++import it.unimi.dsi.fastutil.ints.*;
++import it.unimi.dsi.fastutil.longs.*;
++import it.unimi.dsi.fastutil.objects.*;
++import net.minecraft.server.ChunkCoordIntPair;
++import net.minecraft.server.EntityPlayer;
++
++import javax.annotation.Nullable;
++
++/*
++  base logic note for ticket level per player view distance.
++
++  in the default update queue, only changed minimum levels are updated
++  this poses issues when view distances overlap, as we hit a midpoint barrier
++
++  e.g overlapping levels which do not surpass the peak midpoint
++  p1 has vd of 8, p2 has vd of 2, p1 and p2 have 9 chunks in between
++  p1 1 2 3 4 5 4 3 2 1 p2
++  p1 x x x x x x x y y y
++  x are tickets added by p1, and y are tickets added by p2, we want the action above.
++  in reality however this happens as only chunks before the midpoint are updated by mojang's default queue when p1 moves
++  p1 1 2 3 4 5 4 3 2 1 p2
++  p1 x x x x _ _ _ y y y
++  _ represents unloaded chunks without tickets, thus this default model fails and we must use an algorithm to account for overlaps & all cases efficiently
++
++  using coordinate compression we graph overlapping squares & squares that need action to compressed cells accordingly
++  we extract the corner coordinates of each filled cell that requires an action, loop through the coordinates and add the pos pairs accordingly to our update queue
++
++  we also simultaneously calculate the area map in the same iteration for maximum efficiency
++
++*/
++
++/**
++ * This class calculates a player ticket attached area map, a player tracked area map that excludes chunks with no tickets attached, as well as
++ * as a binary update queue simultaneously, through an efficient although specific coordinate compression algorithm.
++ *
++ * @author slicklibro */
++
++public class PlayerTicketAreaMap {
++
++    //byte constants, these can be anything but 0
++    public static final byte ADD = 1;
++    public static final byte REMOVE = -1;
++    public static final byte EXISTING = -2;
++    public static final byte TRACK = -3;
++    public static final byte UNTRACK = -4;
++    public static final byte OVERLAP = -5;
++    public static final byte TRACK_OVERLAP = -6;
++    public static final byte TRACK_EXISTING = -7;
++
++    private final Int2ObjectMap<Square> squares = new Int2ObjectOpenHashMap<>();
++    private final Int2ObjectMap<Square> trackSquares = new Int2ObjectOpenHashMap<>();
++
++    private Long2ByteMap toUpdate = new Long2ByteOpenHashMap();
++
++    //players mapped to chunks that should be ticketed
++    public final Long2ObjectMap<ObjectSet<EntityPlayer>> areaMap = new Long2ObjectOpenHashMap<>();
```

Comment:
The problem with this is that the `PlayerMobDistanceMap` is capable of doing this mapping already and it's been tested to work. There's no reason to rewrite the utility unless you have good reason.
The reason you might have is that your utility is capable of propagating updates without a backing map, however it's not like the `PlayerMobDistanceMap` is incapable either. The logic is all in the update method.

The downside to using a map that can propagate updates without a backing map is that you're not able to tell efficiently whether a point is previously tracked by another player, which looks pretty critical for using it for per-player view distance.

Reply:
Also noting that `PlayerMobDistanceMap` is far better with memory usage (and easier on the GC) given it uses pooled sets. It can also technically share those sets across maps if a constructor is added.

Reply:
solved in rewrite already... the algorithm performs far better and serves a much more different purpose, also your pooled hash sets can be used

Reply:
I'm interested in those benchmarks